### ISSUE TITLE
pdm pinned wrong package dependencies

### ISSUE THREAD

**Issue by he0119** [pdm-project/pdm#1803]
- [x] I have searched the issue tracker and believe that this is not a duplicate.

**Make sure you run commands with `-v` flag before pasting the output.**

## Steps to reproduce

1. `pdm add git+https://github.com/nonebot/adapter-console.git`

## Actual behavior

got [`textual<0.11.0,>=0.10.1`](https://github.com/he0119/pdm-test/blob/c89ea90fa366e6554ea43b2bd02ab0427d6f8d6e/pdm.lock#L70)

```
[[package]]
name = "nonebot-adapter-console"
version = "0.3.2"
requires_python = ">=3.8,<4.0"
git = "https://github.com/nonebot/adapter-console.git"
revision = "57cbf8c79dbafed986bd6daaeaf9f9628db4a47d"
summary = "console adapter for nonebot2"
dependencies = [
    "nonebot2<3.0.0,>=2.0.0-beta.1",
    "textual<0.11.0,>=0.10.1",
]
```

## Expected behavior

`textual<0.18.0,>=0.17.0`

## Environment Information

```bash
❯ pdm info && pdm info --env
PDM version:
  2.5.0b0
Python Interpreter:
  C:\Users\hmy01\Works\Working\Bot\pdm-test\.venv\Scripts\python.exe (3.11)
Project Root:
  C:/Users/hmy01/Works/Working/Bot/pdm-test
Local Packages:

{
  "implementation_name": "cpython",
  "implementation_version": "3.11.2",
  "os_name": "nt",
  "platform_machine": "AMD64",
  "platform_release": "10",
  "platform_system": "Windows",
  "platform_version": "10.0.22621",
  "python_full_version": "3.11.2",
  "platform_python_implementation": "CPython",
  "python_version": "3.11",
  "sys_platform": "win32"
}
```

**Comment by frostming** [comment]
I can't reproduce locally, does this issue still exist after you clear the cache?

**Comment by he0119** [comment]
> I can't reproduce locally, does this issue still exist after you clear the cache?

It works after clearing the cache.🤣

<https://github.com/he0119/pdm-test/commit/1c9d89a6c9691a890b26589ff4f1f53b0250038a#diff-4dc59eeadc4eee1785e5a8461355721763a4ab9a32b8e3b88def661da9cb497cR95>

**Comment by frostming** [comment]
Did you do any abnormal things to the Github repo, such as modifying the git history. The cache shouldn't be hit if the commit hash changes.

**Comment by he0119** [comment]
The nonebot-adapter-console repo's master branch is protected and the git history cannot be modified.

I checked my project commit history, only found this <https://github.com/nonepkg/nonebot-plugin-all4one/pull/22/commits/d061e5c4240943b1d4c19c9bac4c9cfc4210717d#diff-4dc59eeadc4eee1785e5a8461355721763a4ab9a32b8e3b88def661da9cb497cL435>. the locked revision did not change, but its dependencies changed.

I tried to use the old lock file and do `pdm update`, but I was unable to reproduce the issue.
